middle_end/flambda2/simplify/inlining/inlining_report.ml

```diff
@@ -4,128 +4,660 @@
+            "A key of type call should be associated with an item of type call")
+      t
+
+  let insert_or_update_fundecl ?before_simplify ?after_simplify
+      ~compilation_unit ~dbg ~name ~path ~(cont : item Map.t -> item Map.t) t =
+    let key = dbg, Key.Fundecl name in
+    Map.update key
+      (function
+        | None ->
+          Some
+            (Fundecl
+               { body = cont Map.empty;
+                 before_simplify;
+                 after_simplify;
+                 uid = Uid.create ~compilation_unit path
+               })
+        | Some (Fundecl t) ->
+          let before_simplify =
+            join_decision t.before_simplify before_simplify
+              ~error:
+                "Each function can have at most one associated decision for \
+                 the before simplify pass."
+          in
+          let after_simplify =
+            join_decision t.after_simplify after_simplify
+              ~error:
+                "Each function can have at most one associated decision for \
+                 the after simplify pass."
+          in
+          Some
+            (Fundecl
+               { body = cont t.body;
+                 after_simplify;
+                 before_simplify;
+                 uid = t.uid
+               })
+        | Some (Call _ | Scope _) ->
+          Misc.fatal_errorf
+            "A key of type call should be associated with an item of type call")
+      t
+
+  (* Insert a decision into an inlining tree [t].
+
+     Only decisions rooted in [compilation_unit] are kept. *)
+  let insert ~compilation_unit t (decision : raw_decision) =
+    let rec aux ~cont path =
+      match (path : Inlining_history.Absolute.path) with
+      | Empty -> cont t
+      | Unknown { prev } ->
+        aux prev ~cont:(fun m ->
+            insert_or_update_scope ~scope_type:Unknown ~name:"" ~cont m)
+      | Module { name; prev } ->
+        aux prev ~cont:(fun m ->
+            insert_or_update_scope ~scope_type:Module ~name ~cont m)
+      | Class { name; prev } ->
+        aux prev ~cont:(fun m ->
+            insert_or_update_scope ~scope_type:Class ~name ~cont m)
+      | Function { dbg; name; prev } ->
+        aux prev ~cont:(fun m ->
+            insert_or_update_fundecl ~dbg ~name ~cont ~compilation_unit ~path m)
+      | Call { dbg; callee; prev } ->
+        aux prev ~cont:(fun m -> insert_or_update_call ~dbg ~callee ~cont m)
+      | Inline { prev } -> aux prev ~cont
+    in
+
+    let { path; dbg; pass; decision_with_context } = decision in
+    if Compilation_unit.equal compilation_unit
+         (Inlining_history.Absolute.compilation_unit path)
+    then
+      let path = Inlining_history.Absolute.path path in
+      match path with
+      | Unknown _ -> t
+      | Call { callee; prev; _ } ->
+        aux
+          ~cont:(fun m ->
+            insert_or_update_call ?decision_with_context ~dbg ~callee
+              ~cont:(fun x -> x)
+              m)
+          prev
```

Comment:
I'm pretty well lost at this point :-\ What good does it do to update the call with the identity? Just make sure the map exists even if it's empty?

Reply:
I agree that right now this algorithm is convoluted. The main issue is that when I'm traversing the path of the decision to insert it in the tree, I'm actually traversing it in reverse historical order (the history is represented as a linked list where the inner most elements are the oldest ones), whereas I can only traverse the tree from the root to the childs. To avoid having to explicitly construct a reverse path the function `rebuild` (previously named `cont`) is constructing a function that will given an inlining tree create a new inlining tree where nodes are inserted as needed. It's a bit like what is done with `simplify` in Flambda2.

Reply:
Ah ... okay, essentially you're building up a trie search/update operation inside-out.

Hmm ... I don't think `rebuild` makes it much clearer what's going on. It's easier to think of `insert_inner_node` as simplify saying "apply some function `f` to the node at path `p`." Then it's obvious that the recursive case for a path `p'.k` must be "to the node at `p'`, apply the function 'apply `f` to child `k`.' " That would also make `insert_or_update_generic` easier to understand, since now it's just "apply `f` to child `k` of node `n` (creating the child if necessary)". I think I would call `insert_or_update_generic` something like `insert_or_update_child_generic` so that `insert_inner_node` can be `insert_or_update_descendant`.